File: ticket_office/cs/TicketOffice/Domain/SeatFinder.cs
namespace TicketOffice.Domain;

public class SeatFinder
{
    private List<Seat> _seats;

    public SeatFinder(IEnumerable<Seat> seats)
    {
        _seats = seats.ToList();
    }

    public List<Seat> Find(int count)
    {
        var availaibleSeats = new List<Seat>();
        foreach (var seat in _seats)
        {
            if (seat.IsFree())
            {
                availaibleSeats.Add(seat);
            }
        }
        return availaibleSeats.Take(count).ToList();
    }
}

File: java/ticket_office/src/main/java/fr/arolla/trainreservation/ticket_office/domain/SeatFinder.java
package fr.arolla.trainreservation.ticket_office.domain;

import java.util.List;

public class SeatFinder {
  private final Train train;

  public SeatFinder(Train train) {
    this.train = train;
  }

  public List<SeatID> findSeats(int seatCount) {
    var coach = findBestCoach(seatCount);
    var inCoach = train.seatsInCoach(coach);
    var availableSeatsInCoach = inCoach.filter(Seat::isFree).sorted();
    var toBook = availableSeatsInCoach.limit(seatCount);
    return toBook.map(s -> s.id()).sorted().toList();
  }

  private CoachID findBestCoach(int seatCount) {
    for (var coach : train.getCoaches()) {
      if (train.occupancyForCoachAfterBooking(coach, seatCount) <= 0.7) {
        return coach;
      }
    }
    throw new NotEnoughFreeSeatsException();
  }

}
